src/interpreter/ByteCodeGenerator.cpp

```diff
@@ -88,14 +88,36 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
 }
 
 #ifdef ESCARGOT_DEBUGGER
-void ByteCodeGenerateContext::insertBreakpoint(size_t line, Node* node)
+void ByteCodeGenerateContext::insertBreakpoint(size_t index, Node* node)
 {
     ASSERT(m_breakpointContext != nullptr);
 
-    if (line != 0 && line != m_breakpointContext->m_lastBreakpointLine) {
-        m_breakpointContext->m_breakpointLocations.push_back(Debugger::BreakpointLocation(line, (uint32_t)m_byteCodeBlock->currentCodeSize()));
+    if (index == SIZE_MAX) {
+        return;
+    }
+
+    StringView src = m_codeBlock->asInterpretedCodeBlock()->src();
+    ExtendedNodeLOC sourceElementStart = m_codeBlock->asInterpretedCodeBlock()->functionStart();
+    size_t lastLine = m_breakpointContext->m_lastBreakpointLine;
+
+    if (index > m_breakpointContext->m_lastBreakpointIndex) {
+        for (size_t i = m_breakpointContext->m_lastBreakpointIndex; i < index; i++) {
+            char16_t c = src.charAt(i - sourceElementStart.index);
+            if (EscargotLexer::isLineTerminator(c)) {
+                // skip \r\n
+                if (c == 13 && (i + 1 < index) && src.charAt(i + 1) == 10) {
+                    i++;
+                }
+                m_breakpointContext->m_lastBreakpointLine++;
+            }
+        }
+
+        m_breakpointContext->m_lastBreakpointIndex = index;
+    }
+
+    if (m_breakpointContext->m_lastBreakpointLine == sourceElementStart.line || lastLine != m_breakpointContext->m_lastBreakpointLine) {
```

Comment:
Why is `m_breakpointContext->m_lastBreakpointLine == sourceElementStart.line` added for this if statement?
IMO `lastLine != m_breakpointContext->m_lastBreakpointLine` check would be enough.

Reply:
I've slightly updated this case, bit there is one corner case:

```js
let j = function () {};

j();
```

Whenever you step in `j()` at least one breakpoint is needed in the function. However,  `lastLine` and `m_breakpointContext->m_lastBreakpointLine` would be 0 at the same time, but the breakpoint is needed, thats why the extra `lastLine == 0` check is needed.

Reply:
I'm little bit confused in line start number (correct me if I'm wrong)
AFAIK line location starts with 1 not 0, so the extra check seems unnecessary.

Reply:
That check is still not unnecessary, only the appellation of the variable is a bit misleading. I'm going to rename it, since now it stores the **offset** from the last line/index not the last line/index.

Reply:
I'm still wondering why `lineOffset` is used to calculate the `lineNumber` instead of directly using the `lineNumber`.
For the following corner case,
```js
let j = function () {};
j();
```
When we step inside `j()`, one breakpoint is triggered because every empty function has one hidden return statement in itself. And the breakpoint of the return statement has of course the same line number of `let j` but this breakpoint is inserted during the bytecode generation of `function j `not global code generation. Since `ByteCodeBreakpointContext` is initialized at each bytecode generation process, I think that `lastLine != m_breakpointContext->m_lastBreakpointLine` check would be enough.

Reply:
Let me revise myself a little bit after some investigation. Whenever the bytecodeGeneration of `function j(){}` start the `m_lastBreakpointLineOffset` is zeroed out in the initialization on the new BytecodeGeneratorContext. Yes there is a manually inserted return node, with a specified node location **index**. Then the lookup starts to find the correspoding line for this specified index and because there is no `\r\n` in the empty function body the line number is not increased. So `lastLineOffset` remains zero and the context's lastLineOffset is also remains zero since it is not increased by the lookup. If we don't check for `lastLineOffset != 0` the `lastLineOffset != m_breakpointContext->m_lastBreakpointLineOffset` compares 0 with 0 which results no breakpoint insertion.

Reply:
Thank you for your neat explanation!
Now I have some minor suggestions (you can ignore it if you don't mind)
- How about storing line/index number instead of offset?

Reply:
I also had a version where I stored the line/index directly in the ByteCodeGenerator context, so I'd insist to the currently solution.